File: Homework1/Hw1/Parser.cs
﻿namespace Hw1;

public static class Parser
{
    public static void ParseCalcArguments(string[] args, 
        out double val1, 
        out CalculatorOperation operation, 
        out double val2)
    {
        if (!IsArgLengthSupported(args))
        {
            throw new ArgumentException();
        }

        if (!double.TryParse(args[0], out val1))
        {
            throw new ArgumentException();
        }

        operation = ParseOperation(args[1]);

        if (!double.TryParse(args[2], out val2))
        {
            throw new ArgumentException();
        }
    }

    private static bool IsArgLengthSupported(string[] args) => args.Length == 3;

    private static CalculatorOperation ParseOperation(string arg)
    {
        switch (arg)
        {
            case "+":
                return CalculatorOperation.Plus;
            case "-":
                return CalculatorOperation.Minus;
            default:
                throw new InvalidOperationException();
        }
    }
}

File: Homework4/Hw4/Parser.fs
module Hw4.Parser

open System
open Hw4.Calculator


type CalcOptions = {
    arg1: float
    arg2: float
    operation: CalculatorOperation
}

let isArgLengthSupported (args : string[]) =
    args.Length = 3

let parseOperation (arg : string) =
    match arg with
    | "+" -> CalculatorOperation.Plus
    | "-" -> CalculatorOperation.Minus
    | "*" -> CalculatorOperation.Multiply
    | "/" -> CalculatorOperation.Divide
    | _ -> ArgumentException("Entered incorrectly operation") |> raise
    
let parseCalcArguments(args : string[]) =
    if (isArgLengthSupported args <> true) then ArgumentException("Incorrect number of arguments entered") |> raise

    let (isConverted1, arg1) = Double.TryParse args[0]
    let (isConverted2, arg2) = Double.TryParse args[2]

    match (isConverted1, isConverted2) with
    | (true, true) -> {arg1 = arg1; arg2 = arg2; operation = parseOperation args[1]}
    | _ -> ArgumentException("Incorrect first or second value entered") |> raise